### PR TITLE
Bundle C++ to SDK to artifact that can be used with CMake's fetch_content

### LINKED ISSUE #3759: Provide CMake setup for using the C++ sdk with pre-built rerun_c
We already have rerun_c binaries uploaded on every devlopment build.
With these it should be possible to provide a CMake setup where a user no longer needs to do any rust build

### PR DESCRIPTION
**Pull request by Wumpf**
### What

* Fixes #3759

Change CMakeLists.txt in rerun_cpp to work standalone with add_directory.
When used in this fashion it expects the `RERUN_C_LIB` variable to be set. The default for it points to a local `lib` folder with appropriately named libs.
This in turn gives rise to a very trivial bundling strategy:
* copy src
* copy CMakeLists.txt
* copy libs

This is done in the new `build_and_upload_rerun_cpp` script which on top of the above also takes care of testing the resulting artifact by running the now also compatible minimal example outside of the repo.

Another smaller necessary change is that the rerun.h C header is now copied into the rerun_cpp sdk with a patched version number in the comment. This is primarily so to keep the above process super simple and shave of an additional include directory.

~Draft until PR ci confirmed to be working.~
(PR checks only the linux artifact, main ci checks the full artifact and puts it on the Development Release notes)

With this change, the minimal possible cmake script for a project using Rerun is:
```cmake
project(hello_rerun)

include(FetchContent)
FetchContent_Declare(rerun_sdk URL https://build.rerun.io/commit/f4e5ad6/rerun_cpp.zip DOWNLOAD_EXTRACT_TIMESTAMP YES)
FetchContent_MakeAvailable(rerun_sdk)

add_executable(hello_rerun main.cpp)
set_property(TARGET hello_rerun PROPERTY CXX_STANDARD 20) # Rerun requires C++17, but you can of course use an ever newer version.
target_link_libraries(hello_rerun PRIVATE rerun_sdk)
```
⚠️  **this STILL requires libarrow to be installed somewhere where cmake can find it** ⚠️ 

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3875) (if applicable)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/3875)
- [Docs preview](https://rerun.io/preview/9ddbc213c8f4bd2ea288590f9ab2f2b9d48ae552/docs) 
- [Examples preview](https://rerun.io/preview/9ddbc213c8f4bd2ea288590f9ab2f2b9d48ae552/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -68,7 +68,12 @@
     "python.analysis.extraPaths": [
         "rerun_py/rerun_sdk"
     ],
+    "C_Cpp.default.configurationProvider": "ms-vscode.cmake-tools", // Use cmake-tools to grab configs.
     "cmake.buildDirectory": "${workspaceRoot}/build/",
+    "cmake.buildTask": true, // Use a terminal instead of the output panel for builds, it won't show color.
```
Comment (emilk) [inline .vscode/settings.json:73]: what part of the sentence does "it" refer to here?

File: crates/rerun_c/src/rerun.h [hunk 2]
```diff
@@ -24,12 +29,12 @@ enum {
...
 /// globally available recording stream for recordings.
 /// (i.e. thread-local first, then global scope)
-#define RERUN_REC_STREAM_CURRENT_RECORDING ((rr_recording_stream)0xFFFFFFFF)
+#define RERUN_REC_STREAM_CURRENT_RECORDING (0xFFFFFFFF)
 
 /// Special value for `rr_recording_stream` methods to indicate the most appropriate
 /// globally available recording stream for blueprints.
 /// (i.e. thread-local first, then global scope)
-#define RERUN_REC_STREAM_CURRENT_BLUEPRINT ((rr_recording_stream)0xFFFFFFFE)
+#define RERUN_REC_STREAM_CURRENT_BLUEPRINT (0xFFFFFFFE)
```
Comment (emilk) [inline crates/rerun_c/src/rerun.h:37]: unnecessary paranthesis

File: CMakeLists.txt [hunk 3]
```diff
@@ -55,6 +55,16 @@ if(NOT DEFINED CMAKE_CXX_STANDARD)
     set(CMAKE_CXX_STANDARD 17)
 endif()
 
+# Signal to all our build scripts that we're inside the Rerun repository.
+set(RERUN_REPOSITORY YES)
+
+# Set version number.
+# Read it from the Rust toml with a regex so we don't have another place to maintain!
```
Comment (emilk) [inline CMakeLists.txt:62]: :+1:

File: examples/cpp/minimal/CMakeLists.txt [hunk 4]
```diff
@@ -1,6 +1,26 @@
 cmake_minimum_required(VERSION 3.16)
 
-add_executable(example_minimal main.cpp)
+# If you use the example outside of the Rerun SDK you need to specify
+# where the rerun_c build is to be found.
```
Comment (emilk) [inline examples/cpp/minimal/CMakeLists.txt:4]: Ok - how do I do that? The next few lines seems unrelated.

File: examples/cpp/minimal/CMakeLists.txt [hunk 5]
```diff
@@ -1,6 +1,26 @@
...
+
+    # Set the path to the rerun_c build.
+    set(RERUN_CPP_URL CACHE STRING "URL to the rerun_cpp zip.")
+
+    # Download the rerun_sdk
+    include(FetchContent)
+    FetchContent_Declare(rerun_sdk DOWNLOAD_EXTRACT_TIMESTAMP YES URL ${RERUN_CPP_URL})
+    FetchContent_MakeAvailable(rerun_sdk)
+
+    # Rerun requires C++17, but you can use an evern newer version.
```
Comment (emilk) [inline examples/cpp/minimal/CMakeLists.txt:21]:
```suggestion
    # Rerun requires at least C++17
```

File: rerun_cpp/CMakeLists.txt [hunk 6]
```diff
@@ -11,64 +11,115 @@ file(GLOB_RECURSE rerun_sdk_SRC CONFIGURE_DEPENDS
 
 add_library(rerun_sdk ${rerun_sdk_SRC})
 
-set_default_warning_settings(rerun_sdk)
+# Rerun needs C++17.
```
Comment (emilk) [inline rerun_cpp/CMakeLists.txt:14]:
```suggestion
# Rerun needs at least C++17.
```

File: rerun_cpp/CMakeLists.txt [hunk 7]
```diff
@@ -11,64 +11,115 @@ file(GLOB_RECURSE rerun_sdk_SRC CONFIGURE_DEPENDS
...
+set_property(TARGET rerun_sdk PROPERTY CXX_STANDARD 17)
 
 # ------------------------------------------------------------------------------
+# Rerun development setup only:
+if(DEFINED RERUN_REPOSITORY)
+    # execute process doesn't fail if the process fails.
+    # `COMMAND_ERROR_IS_FATAL ANY` parameter fixes this but is only available in CMake 3.19
+    # which isn't default on Ubuntu LTS as of writing, causing unnecessary friction.
+    # TODO(andreas): use add_custom_command instead so this runs at build time! https://cmake.org/cmake/help/latest/command/add_custom_command.html#command:add_custom_command
+    execute_process(COMMAND cargo build --release -p re_types RESULT_VARIABLE ret) # Generates most of the C++ source files
```
Comment (emilk) [inline rerun_cpp/CMakeLists.txt:24]:
```suggestion
    execute_process(COMMAND just codegen RESULT_VARIABLE ret) # Generates most of the C++ source files
```

File: rerun_cpp/CMakeLists.txt [hunk 8]
```diff
@@ -11,64 +11,115 @@ file(GLOB_RECURSE rerun_sdk_SRC CONFIGURE_DEPENDS
...
+        set(RERUN_C_LIB_DEFAULT ${CMAKE_CURRENT_SOURCE_DIR}/../target/release/rerun_c.lib)
+    else()
+        message(FATAL_ERROR "Unsupported platform.")
+    endif()
+
+    # Set very strict warning settings when we're testing the SDK.
+    # We don't want to force this on any user!
+    set_default_warning_settings(rerun_sdk)
+
+    # Put Rerun.h into the same place where it's on a user's machine and apply CMake variables like version number.
```
Comment (emilk) [inline rerun_cpp/CMakeLists.txt:51]:
```suggestion
    # Put `rerun.h` into the same place where it's on a user's machine and apply CMake variables like version number.
```

File: scripts/ci/bundle_and_upload_rerun_cpp.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+def upload_rerun_cpp(rerun_zip: str, git_hash: str) -> None:
+    logging.info("Uploading to gcloud...")
+
+    gcs = storage.Client("rerun-open")
+    bucket = gcs.bucket("rerun-builds")
+    destination = bucket.blob(f"commit/{git_hash}/rerun_cpp.zip")
+    destination.content_type = "application/zip"
+    destination.upload_from_filename(rerun_zip)
+    logging.info(f"Uploaded to https://build.rerun.io/commit/{git_hash}/rerun_cpp.zip")
```
Comment (emilk) [inline scripts/ci/bundle_and_upload_rerun_cpp.py:54]: nitpick, but I think we should call it `rerun_cpp_sdk.zip` for extra explicitness

File: scripts/ci/bundle_and_upload_rerun_cpp.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+
+def test_rerun_cpp(git_hash: str) -> None:
+    logging.info("Testing uploaded artifact...")
```
Comment (emilk) [inline scripts/ci/bundle_and_upload_rerun_cpp.py:58]:
```suggestion
    logging.info("Testing uploaded artifact…")
```

File: scripts/ci/bundle_and_upload_rerun_cpp.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+) -> None:
+    print(f"> {subprocess.list2cmdline(args)}")
+    result = subprocess.run(args, env=env, cwd=cwd, timeout=timeout, check=False, capture_output=True, text=True)
+    assert (
+        result.returncode == 0
+    ), f"{subprocess.list2cmdline(args)} failed with exit-code {result.returncode}. Output:\n{result.stdout}\n{result.stderr}"
+
+
+def download_rerun_c(target_dir: str, git_hash: str, platform_filter: str = None) -> None:
+    logging.info("Downloading rerun_c...")
```
Comment (emilk) [inline scripts/ci/bundle_and_upload_rerun_cpp.py:27]:
```suggestion
    logging.info("Downloading rerun_c…")
```

**Review by emilk — APPROVED** [review]
This is awesome!

**Review by teh-cmc — APPROVED** [review]
Looks fantastic.

I just wish we did the same for arrow-cpp now!

File: .vscode/settings.json (hunk 1 above)
Comment (Wumpf) [reply .vscode/settings.json:73]: the `buildTask` 😄

File: rerun_cpp/CMakeLists.txt (hunk 7 above)
Comment (Wumpf) [reply rerun_cpp/CMakeLists.txt:24]: It makes sense, but I'm unhappy about this requiring more dependencies for people coming in to build cpp themselves

File: scripts/ci/bundle_and_upload_rerun_cpp.py (hunk 10 above)
Comment (Wumpf) [reply scripts/ci/bundle_and_upload_rerun_cpp.py:58]: I have sinned!

File: rerun_cpp/CMakeLists.txt (hunk 7 above)
Comment (Wumpf) [reply rerun_cpp/CMakeLists.txt:24]: actually I think now we should just remove this line. Codegen should *always* be explicit now

**Comment by Wumpf** [comment]
> I just wish we did the same for arrow-cpp now!

my gut feeling tells me this will be **way** harder, but I haven't tried!
